test/suite/test_cursor17.py

```diff
@@ -0,0 +1,120 @@
+# cursor traversal mechanism to skip pages where all records on the page are deleted with a
+# tombstone visible to the current transaction.
+class test_cursor17(wttest.WiredTigerTestCase):
+    conn_config = 'cache_size=50MB,statistics=(all)'
+    session_config = 'isolation=snapshot'
+
+    def get_stat(self, stat, uri):
+        stat_string = 'statistics:'
+        if (uri):
+            stat_string += uri
+        stat_cursor = self.session.open_cursor(stat_string)
+        val = stat_cursor[stat][2]
+        stat_cursor.close()
+        return val
+
+    def test_cursor_skip_pages(self):
+        uri = 'table:test_cursor17'
+        create_params = 'key_format=i,value_format=S'
+        self.session.create(uri, create_params)
+
+        value1 = 'a' * 500
+        value2 = 'b' * 500
+        total_keys = 40000
+
+        # Keep the oldest and the stable timestamp pinned.
+        self.conn.set_timestamp('oldest_timestamp=' + timestamp_str(1))
+        self.conn.set_timestamp('stable_timestamp=' + timestamp_str(2))
+        cursor = self.session.open_cursor(uri)
+
+        commit_timestamp = 3
+
+        # Insert few thousand key-value pairs.
+        for key in range(total_keys):
+            self.session.begin_transaction()
+            cursor[key] = value1
+            self.session.commit_transaction('commit_timestamp=' + timestamp_str(commit_timestamp))
+            commit_timestamp += 1
+
+        # Delete everything on the table except for the first and the last KV pair.
+        for key in range(1, total_keys - 1):
+            self.session.begin_transaction()
+            cursor.set_key(key)
+            self.assertEqual(cursor.remove(),0)
+            self.session.commit_transaction('commit_timestamp=' + timestamp_str(commit_timestamp))
+            commit_timestamp += 1
+
+        # Take a checkpoint to reconcile the pages.
+        self.session.checkpoint()
+
+        self.session.begin_transaction('read_timestamp=' + timestamp_str(commit_timestamp))
+        # Position the cursor on the first record.
+        cursor.set_key(0)
+        self.assertEqual(cursor.search(), 0)
+        # This should move the cursor to the last record.
+        self.assertEqual(cursor.next(), 0)
+        self.assertEqual(cursor.get_key(), total_keys - 1)
+
+        # Check if we skipped any pages while moving the cursor.
+        #
+        # We calculate the number of pages we expect to skip based on the total number of leaf pages
+        # reported in the WT stats. We subtract 2 from the count of leaf pages in the table and test
+        # that we atleast skipped 80% of the expected number of pages.
+
+        leaf_pages_in_table = self.get_stat(stat.dsrc.btree_row_leaf, uri)
+        expected_pages_skipped = ((leaf_pages_in_table - 2) * 8) // 10
+        skipped_pages = self.get_stat(stat.conn.cursor_next_skip_page_count, None)
+        self.assertGreater(skipped_pages, expected_pages_skipped)
```

Comment:
Is this test reliable? I think if we fail to lock the page, the number of skipped pages would be less than expected, is that correct?

Reply:
The cache size used for this test is 50MB and table size is around 24MB. Therefore, I expect minimal eviction activity once I have performed a checkpoint. I have kept a 20% leeway in case there are pages that we somehow are unable to skip.

Reply:
I think we only need to assert the page skip happens and that should be safer? I don't understand why we need to assert that we skip a certain amount of pages to test the feature is working. I think this is not a performance test, isn't it?

Reply:
The intent is to test the optimization. Arguably this may not be the best or most reliable approach. I believe there is follow-on work to add this scenario to our `workgen` performance testing - @hbokhari do you feel that would be a more appropriate place to add this testing?

Reply:
I believe performance testing with `workgen` would indicate if this optimization is giving us any benefit with certain workloads. Here, I am functionally testing that optimization is working as intended ,i.e., it is able to skip pages given certain conditions such as larger than table cache and a clean tree in cache (no eviction activity). If the assumption here that optimization should skip all pages between first and the last page becomes invalid, we might want to go back and review the optimization.

Having said that, if a majority of reviewer feels the test is too strict, I am happy to remove the constraint.

Reply:
@hbokhari I'm happy to trust your judgement on this as you've been working on the issue for several months.